src/Features/CSharp/Portable/UsePatternCombinators/AnalyzedPattern.cs

```diff
@@ -0,0 +1,124 @@
+                => TypeSyntax = expression;
+        }
+
+        /// <summary>
+        /// Represents a source-pattern, constructed from C# patterns
+        /// </summary>
+        internal sealed class Source : AnalyzedPattern
+        {
+            public readonly PatternSyntax PatternSyntax;
+
+            public Source(PatternSyntax patternSyntax)
+                => PatternSyntax = patternSyntax;
+        }
+
+        /// <summary>
+        /// Represents a constant-pattern, constructed from an equality check
+        /// </summary>
+        internal sealed class Constant : AnalyzedPattern
+        {
+            public readonly ExpressionSyntax ExpressionSyntax;
+
+            public Constant(ExpressionSyntax expression)
+                => ExpressionSyntax = expression;
+        }
+
+        /// <summary>
+        /// Represents a relational-pattern, constructed from relational operators
+        /// </summary>
+        internal sealed class Relational : AnalyzedPattern
+        {
+            public readonly BinaryOperatorKind OperatorKind;
+            public readonly ExpressionSyntax Value;
+
+            public Relational(BinaryOperatorKind operatorKind, ExpressionSyntax value)
+            {
+                OperatorKind = operatorKind;
+                Value = value;
+            }
+        }
+
+        /// <summary>
+        /// Represents an and/or pattern, constructed from a logical and/or expression.
+        /// </summary>
+        internal sealed class Binary : AnalyzedPattern
+        {
+            public readonly AnalyzedPattern Left;
+            public readonly AnalyzedPattern Right;
+            public readonly bool IsDisjunctive;
+            public readonly SyntaxToken Token;
+
+            private Binary(AnalyzedPattern leftPattern, AnalyzedPattern rightPattern, bool isDisjunctive, SyntaxToken token)
+            {
+                Left = leftPattern;
+                Right = rightPattern;
+                IsDisjunctive = isDisjunctive;
+                Token = token;
+            }
+
+            public static AnalyzedPattern Create(AnalyzedPattern leftPattern, AnalyzedPattern rightPattern, bool isDisjunctive, SyntaxToken token)
+            {
+                return (leftPattern, rightPattern) switch
+                {
+                    (Not left, Not right) => Not.Create(new Binary(left.Pattern, right.Pattern, !isDisjunctive, token)),
+                    _ => new Binary(leftPattern, rightPattern, isDisjunctive, token)
+                };
+            }
+        }
+
+        /// <summary>
+        /// Represents a not-pattern, constructed from inequality check or a logical-not expression.
+        /// </summary>
+        internal sealed class Not : AnalyzedPattern
+        {
+            public readonly AnalyzedPattern Pattern;
+
+            private Not(AnalyzedPattern pattern) => Pattern = pattern;
+
+            private static BinaryOperatorKind Negate(BinaryOperatorKind kind) => kind switch
```

Comment:
so we already have a negate helper for patterns.  i dont' love that we now have two.

Reply:
There's lots of code that could be shared with if-to-switch. I just didn't go thought that yet.
I'd prefer this to be done separately.